zulip_bots/zulip_bots/bots/googletranslate/googletranslate.py

```diff
@@ -0,0 +1,104 @@
+                                                  message['sender_full_name'],
+                                                  self.supported_languages)
+        bot_handler.send_reply(message, bot_response)
+
+api_url = 'https://translation.googleapis.com/language/translate/v2'
+
+help_text = '''
+Google translate bot
+Please format your message like:
+`@-mention "<text_to_translate>" <target-language> <source-language(optional)>`
+Visit [here](https://cloud.google.com/translate/docs/languages) for all languages
+'''
+
+language_not_found_text = 'Language not found. Visit [here](https://cloud.google.com/translate/docs/languages) for all languages'
+
+def get_supported_languages(key):
+    parameters = {'key': key, 'target': 'en'}
+    response = requests.get(api_url + '/languages', params = parameters)
+    if response.status_code == requests.codes.ok:
+        languages = response.json()['data']['languages']
+        return {lang['name'].lower(): lang['language'].lower() for lang in languages}
+    raise TranslateError(response.json()['error']['message'])
+
+class TranslateError(Exception):
+    pass
+
+def translate(text_to_translate, key, dest, src):
+    parameters = {'q': text_to_translate, 'target': dest, 'key': key}
+    if src != '':
+        parameters.update({'source': src})
+    response = requests.post(api_url, params=parameters)
+    if response.status_code == requests.codes.ok:
+        return response.json()['data']['translations'][0]['translatedText']
+    raise TranslateError(response.json()['error']['message'])
+
+def check_language(language, all_languages):
+    if language.lower() not in all_languages.values():
+        if language.lower() not in all_languages.keys():
+            return ''
+        language = all_languages[language.lower()]
+    return language
+
+def get_translate_bot_response(message_content, config_file, author, all_languages):
+    message_content = message_content.strip()
+    if message_content == 'help' or message_content is None or not message_content.startswith('"'):
+        # Return the help message
+        return help_text
+    split_text = message_content.rsplit('" ', 1)
+    if len(split_text) == 1:
+        return help_text
+    split_text += split_text.pop(1).split(' ')
+    if len(split_text) == 2:
+        # There is no source language
+        split_text.append("")
+    if len(split_text) != 3:
+        return help_text
+    (text_to_translate, target_language, source_language) = split_text
+    text_to_translate = text_to_translate[1:]
+    target_language = check_language(target_language, all_languages)
+    if target_language == '':
+        return language_not_found_text
+    if source_language != '':
+        source_language = check_language(source_language, all_languages)
+        if source_language == '':
+            return language_not_found_text
+    try:
+        translated_text = translate(text_to_translate, config_file['key'], target_language, source_language)
```

Comment:
This is so much clearer to read, and I hope you agree makes the rest of the function easier to understand too?